File: firmware/src/comm/Protocol.cpp
#include "Protocol.h"

#include <string.h>

namespace TwinProtocol {

uint16_t crc16CcittFalse(const uint8_t* data, size_t length) {
  uint16_t crc = 0xFFFF;
  for (size_t i = 0; i < length; ++i) {
    crc ^= static_cast<uint16_t>(data[i]) << 8;
    for (int bit = 0; bit < 8; ++bit) {
      if (crc & 0x8000) {
        crc = static_cast<uint16_t>((crc << 1) ^ 0x1021);
      } else {
        crc <<= 1;
      }
    }
  }
  return crc;
}

static void writeU16LE(uint8_t* out, uint16_t value) {
  out[0] = static_cast<uint8_t>(value & 0xFF);
  out[1] = static_cast<uint8_t>((value >> 8) & 0xFF);
}

static void writeU64LE(uint8_t* out, uint64_t value) {
  for (int i = 0; i < 8; ++i) {
    out[i] = static_cast<uint8_t>((value >> (8 * i)) & 0xFF);
  }
}

static void writeF32LE(uint8_t* out, float value) {
  memcpy(out, &value, sizeof(float));
}

size_t encodeTelemetry(const TelemetryPacket& packet, uint8_t* out, size_t outSize) {
  constexpr uint16_t payloadLen = 24;
  constexpr size_t frameLen = kHeaderSize + payloadLen + kCrcSize;
  if (outSize < frameLen) {
    return 0;
  }

  out[0] = kSync0;
  out[1] = kSync1;
  out[2] = kVersion;
  out[3] = kTypeTelemetry;
  writeU16LE(out + 4, packet.sequence);
  writeU16LE(out + 6, payloadLen);
  writeU16LE(out + 8, 0);

  uint8_t* payload = out + kHeaderSize;
  writeU64LE(payload + 0, packet.timestampUs);
  writeF32LE(payload + 8, packet.dutyA);
  writeF32LE(payload + 12, packet.dutyB);
  writeF32LE(payload + 16, packet.dutyC);
  writeF32LE(payload + 20, packet.vdc);

  const uint16_t crc = crc16CcittFalse(out + 2, kHeaderSize - 2 + payloadLen);
  writeU16LE(out + kHeaderSize + payloadLen, crc);
  return frameLen;
}

size_t encodeTwinState(const TwinStatePacket& packet, uint8_t* out, size_t outSize) {
  constexpr uint16_t payloadLen = 20;
  constexpr size_t frameLen = kHeaderSize + payloadLen + kCrcSize;
  if (outSize < frameLen) {
    return 0;
  }

  out[0] = kSync0;
  out[1] = kSync1;
  out[2] = kVersion;
  out[3] = kTypeTwinState;
  writeU16LE(out + 4, packet.sequence);
  writeU16LE(out + 6, payloadLen);
  writeU16LE(out + 8, 0);

  uint8_t* payload = out + kHeaderSize;
  writeF32LE(payload + 0, packet.ia);
  writeF32LE(payload + 4, packet.ib);
  writeF32LE(payload + 8, packet.ic);
  writeF32LE(payload + 12, packet.rotorAngle);
  writeF32LE(payload + 16, packet.rotorSpeed);

  const uint16_t crc = crc16CcittFalse(out + 2, kHeaderSize - 2 + payloadLen);
  writeU16LE(out + kHeaderSize + payloadLen, crc);
  return frameLen;
}

FrameParser::FrameParser() { reset(); }

void FrameParser::reset() {
  length_ = 0;
  memset(buffer_, 0, sizeof(buffer_));
}

void FrameParser::consumeByte(uint8_t byte) {
  if (length_ < sizeof(buffer_)) {
    buffer_[length_++] = byte;
  } else {
    memmove(buffer_, buffer_ + 1, sizeof(buffer_) - 1);
    buffer_[sizeof(buffer_) - 1] = byte;
    length_ = sizeof(buffer_);
  }
}

void FrameParser::resync() {
  if (length_ == 0) {
    return;
  }
  memmove(buffer_, buffer_ + 1, length_ - 1);
  --length_;
}

bool FrameParser::tryParseFrame(Result& result) {
  while (length_ >= 2 && !(buffer_[0] == kSync0 && buffer_[1] == kSync1)) {
    resync();
  }

  if (length_ < kHeaderSize) {
    return false;
  }

  const uint8_t version = buffer_[2];
  const uint8_t type = buffer_[3];
  const uint16_t payloadLen = static_cast<uint16_t>(buffer_[6]) |
                              (static_cast<uint16_t>(buffer_[7]) << 8);
  const size_t frameLen = kHeaderSize + payloadLen + kCrcSize;

  if (version != kVersion) {
    resync();
    return false;
  }

  if (type != kTypeTelemetry && type != kTypeTwinState) {
    resync();
    return false;
  }

  if ((type == kTypeTelemetry && payloadLen != 24) ||
      (type == kTypeTwinState && payloadLen != 20)) {
    resync();
    return false;
  }

  if (length_ < frameLen) {
    return false;
  }

  const uint16_t receivedCrc = static_cast<uint16_t>(buffer_[frameLen - 2]) |
                               (static_cast<uint16_t>(buffer_[frameLen - 1]) << 8);
  const uint16_t calculatedCrc = crc16CcittFalse(buffer_ + 2, kHeaderSize - 2 + payloadLen);
  if (receivedCrc != calculatedCrc) {
    resync();
    return false;
  }

  const uint8_t* payload = buffer_ + kHeaderSize;
  result.valid = true;
  result.type = type;
  result.telemetry.sequence = static_cast<uint16_t>(buffer_[4]) | (static_cast<uint16_t>(buffer_[5]) << 8);
  result.twinState.sequence = result.telemetry.sequence;

  if (type == kTypeTelemetry) {
    memcpy(&result.telemetry.timestampUs, payload + 0, sizeof(uint64_t));
    memcpy(&result.telemetry.dutyA, payload + 8, sizeof(float));
    memcpy(&result.telemetry.dutyB, payload + 12, sizeof(float));
    memcpy(&result.telemetry.dutyC, payload + 16, sizeof(float));
    memcpy(&result.telemetry.vdc, payload + 20, sizeof(float));
  } else {
    memcpy(&result.twinState.ia, payload + 0, sizeof(float));
    memcpy(&result.twinState.ib, payload + 4, sizeof(float));
    memcpy(&result.twinState.ic, payload + 8, sizeof(float));
    memcpy(&result.twinState.rotorAngle, payload + 12, sizeof(float));
    memcpy(&result.twinState.rotorSpeed, payload + 16, sizeof(float));
  }

  memmove(buffer_, buffer_ + frameLen, length_ - frameLen);
  length_ -= frameLen;
  return true;
}

bool FrameParser::feed(uint8_t byte, Result& result) {
  consumeByte(byte);
  return tryParseFrame(result);
}

}  // namespace TwinProtocol

File: backend/app/serial/protocol.py
from __future__ import annotations

from dataclasses import dataclass
from enum import IntEnum
import struct
from typing import Optional, Tuple


SYNC = b"\xAA\x55"
VERSION = 1
HEADER_SIZE = 10
CRC_SIZE = 2
MAX_FRAME_SIZE = 64


class PacketType(IntEnum):
    TELEMETRY = 0x01
    TWIN_STATE = 0x02


class ProtocolError(Exception):
    """Base class for protocol errors."""


class PacketValidationError(ProtocolError):
    """Raised when a frame fails header or CRC validation."""


class IncompleteFrameError(ProtocolError):
    """Raised when more bytes are needed to complete a frame."""


def crc16_ccitt_false(data: bytes, initial: int = 0xFFFF) -> int:
    crc = initial & 0xFFFF
    for byte in data:
        crc ^= byte << 8
        for _ in range(8):
            if crc & 0x8000:
                crc = ((crc << 1) ^ 0x1021) & 0xFFFF
            else:
                crc = (crc << 1) & 0xFFFF
    return crc


@dataclass(frozen=True)
class TelemetryPacket:
    sequence: int
    timestamp_us: int
    duty_a: float
    duty_b: float
    duty_c: float
    vdc: float


@dataclass(frozen=True)
class TwinStatePacket:
    sequence: int
    timestamp_us: int
    ia: float
    ib: float
    ic: float
    rotor_speed: float
    rotor_angle: float
    electrical_angle: float
    torque: float
    temperature: float
    hall: int
    encoder: int
    fault_flags: int


def _pack_header(packet_type: PacketType, sequence: int, payload_len: int) -> bytes:
    return struct.pack("<2sBBHHH", SYNC, VERSION, int(packet_type), sequence & 0xFFFF, payload_len, 0)


def _unpack_header(frame: bytes) -> Tuple[int, int, int, int, int]:
    sync, version, packet_type, sequence, payload_len, reserved = struct.unpack("<2sBBHHH", frame[:HEADER_SIZE])
    if sync != SYNC:
        raise PacketValidationError("invalid sync bytes")
    if version != VERSION:
        raise PacketValidationError(f"unsupported protocol version: {version}")
    if reserved != 0:
        raise PacketValidationError("reserved header field must be zero")
    return packet_type, sequence, payload_len, version, reserved


def encode_telemetry(packet: TelemetryPacket) -> bytes:
    payload = struct.pack(
        "<Qffff",
        packet.timestamp_us,
        packet.duty_a,
        packet.duty_b,
        packet.duty_c,
        packet.vdc,
    )
    header = _pack_header(PacketType.TELEMETRY, packet.sequence, len(payload))
    crc = crc16_ccitt_false(header[2:] + payload)
    return header + payload + struct.pack("<H", crc)


def encode_twin_state(packet: TwinStatePacket) -> bytes:
    payload = struct.pack(
        "<QffffffffBHI",
        packet.timestamp_us,
        packet.ia,
        packet.ib,
        packet.ic,
        packet.rotor_speed,
        packet.rotor_angle,
        packet.electrical_angle,
        packet.torque,
        packet.temperature,
        packet.hall,
        packet.encoder,
        packet.fault_flags,
    )
    header = _pack_header(PacketType.TWIN_STATE, packet.sequence, len(payload))
    crc = crc16_ccitt_false(header[2:] + payload)
    return header + payload + struct.pack("<H", crc)


def _validate_and_split(frame: bytes) -> Tuple[PacketType, int, bytes]:
    if len(frame) < HEADER_SIZE + CRC_SIZE:
        raise IncompleteFrameError("frame is too short")
    if frame[:2] != SYNC:
        raise PacketValidationError("invalid sync bytes")

    packet_type_raw, sequence, payload_len, _, _ = _unpack_header(frame)
    try:
        packet_type = PacketType(packet_type_raw)
    except ValueError as exc:
        raise PacketValidationError(f"unknown packet type: {packet_type_raw}") from exc

    expected_len = HEADER_SIZE + payload_len + CRC_SIZE
    if len(frame) != expected_len:
        raise PacketValidationError(
            f"length mismatch: expected {expected_len} bytes, got {len(frame)} bytes"
        )

    payload = frame[HEADER_SIZE:HEADER_SIZE + payload_len]
    received_crc = struct.unpack("<H", frame[-CRC_SIZE:])[0]
    calculated_crc = crc16_ccitt_false(frame[2:-CRC_SIZE])
    if received_crc != calculated_crc:
        raise PacketValidationError("CRC mismatch")

    if packet_type == PacketType.TELEMETRY and payload_len != 24:
        raise PacketValidationError("invalid telemetry payload length")
    if packet_type == PacketType.TWIN_STATE and payload_len != 47:
        raise PacketValidationError("invalid twin state payload length")

    return packet_type, sequence, payload


def decode_frame(frame: bytes) -> TelemetryPacket | TwinStatePacket:
    packet_type, sequence, payload = _validate_and_split(frame)

    if packet_type == PacketType.TELEMETRY:
        timestamp_us, duty_a, duty_b, duty_c, vdc = struct.unpack("<Qffff", payload)
        return TelemetryPacket(
            sequence=sequence,
            timestamp_us=timestamp_us,
            duty_a=duty_a,
            duty_b=duty_b,
            duty_c=duty_c,
            vdc=vdc,
        )

    (
        timestamp_us,
        ia,
        ib,
        ic,
        rotor_speed,
        rotor_angle,
        electrical_angle,
        torque,
        temperature,
        hall,
        encoder,
        fault_flags,
    ) = struct.unpack("<QffffffffBHI", payload)
    return TwinStatePacket(
        sequence=sequence,
        timestamp_us=timestamp_us,
        ia=ia,
        ib=ib,
        ic=ic,
        rotor_speed=rotor_speed,
        rotor_angle=rotor_angle,
        electrical_angle=electrical_angle,
        torque=torque,
        temperature=temperature,
        hall=hall,
        encoder=encoder,
        fault_flags=fault_flags,
    )


class FrameBuffer:
    """Incremental stream parser for a byte-oriented serial link."""

    def __init__(self) -> None:
        self._buffer = bytearray()

    def feed(self, data: bytes) -> list[TelemetryPacket | TwinStatePacket]:
        self._buffer.extend(data)
        packets: list[TelemetryPacket | TwinStatePacket] = []

        while True:
            frame = self._extract_next_frame()
            if frame is None:
                break
            try:
                packets.append(decode_frame(frame))
            except ProtocolError:
                continue

        return packets

    def _extract_next_frame(self) -> Optional[bytes]:
        while len(self._buffer) >= 2 and self._buffer[:2] != SYNC:
            del self._buffer[0]

        if len(self._buffer) < HEADER_SIZE:
            return None

        try:
            _, _, payload_len, _, _ = _unpack_header(bytes(self._buffer[:HEADER_SIZE]))
        except PacketValidationError:
            del self._buffer[0]
            return None

        if payload_len > MAX_FRAME_SIZE:
            del self._buffer[0]
            return None

        frame_len = HEADER_SIZE + payload_len + CRC_SIZE
        if len(self._buffer) < frame_len:
            return None

        frame = bytes(self._buffer[:frame_len])
        del self._buffer[:frame_len]
        return frame
